safehttp/form.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"mime/multipart"
+	"strconv"
+)
+
+// Form contains parsed data either from URL field's query parameters or from
+// form parameters, part of the body of POST, PATCH or PUT requests that are not
+// multipart requests. The fields are only available after parsing the form,
+// through getter functions that specify the type.
+type Form struct {
+	values map[string][]string
+	err    error
+}
+
+// MultipartForm extends the Form structure to define a POST, PATCH or PUT
+// request that is multipart. Its fields are only available after parsing the
+// form, through getter functions that specify the type.
+type MultipartForm struct {
+	*Form
```

Comment:
why a pointer here? MultipartForm will always be used by ref, so we can embed a value here.

Reply:
Good question... resolved in [d511d47](url)